packages/core/src/modules/oob/OutOfBandModule.ts

```diff
@@ -0,0 +1,297 @@
+        connectionRecord = await this.createConnection(outOfBandMessage, autoAccept)
+      }
+
+      if (messages) {
+        this.logger.debug('Out of band message contains request messages.')
+        if (!connectionRecord.isReady) {
+          // Wait until the connecion is ready and then pass the messages to the agent for further processing
+          this.connectionsModule
+            .returnWhenIsConnected(connectionRecord.id)
+            .then((c) => {
+              const connectionServices = c.theirDidDoc?.didCommServices
+              this.emitMessages(connectionServices, messages)
+            })
+            .catch((error) => {
+              if (error instanceof EmptyError) {
+                this.logger.warn(
+                  `Agent unsubscribed before connection got into ${ConnectionState.Complete} state`,
+                  error
+                )
+              } else {
+                this.logger.error('Promise waiting for the connection to be complete failed.', error)
+              }
+            })
+        } else {
+          const connectionServices = connectionRecord.theirDidDoc?.didCommServices
+          this.emitMessages(connectionServices, messages)
+        }
+      }
+
+      return connectionRecord
+    } else if (messages) {
+      this.emitMessages(services, messages)
+    }
+  }
+
+  private areHandshakeProtocolsSupported(handshakeProtocols: string[]) {
+    return this.getSupportedHandshakeProtocols().some((p) => handshakeProtocols.includes(p))
+  }
+
+  private getSupportedHandshakeProtocols() {
+    const handshakeMessageFamilies = ['https://didcomm.org/didexchange', 'https://didcomm.org/connections']
+    const handshakeProtocols = this.dispatcher.filterSupportedProtocolsByMessageFamilies(handshakeMessageFamilies)
+
+    if (handshakeProtocols.length === 0) {
+      throw new AriesFrameworkError('There is no handshake protocol supported. Agent can not create a connection.')
+    }
+
+    return handshakeProtocols
+  }
+
+  private async findExistingConnection(services: DidCommService[]) {
+    for (const service of services) {
+      for (const recipientKey of service.recipientKeys) {
+        // TODO Encode the key and endpoint of the service block in a Peer DID numalgo 2 and using that DID instead of a service block
+        const existingConnection = await this.connectionsModule.findByTheirKey(recipientKey)
+        return existingConnection
+      }
+    }
+  }
+
+  private async createConnection(outOfBandMessage: OutOfBandMessage, autoAccept: boolean) {
+    const { services, label } = outOfBandMessage
+    const invitation = new ConnectionInvitationMessage({
+      label: label || '',
+      ...services[0],
+    })
+    const connectionRecord = await this.connectionsModule.receiveInvitation(invitation, {
+      autoAcceptConnection: autoAccept,
+    })
+    return connectionRecord
+  }
+
+  private emitMessages(services: DidCommService[] | undefined, messages: PlaintextMessage[]) {
```

Comment:
Having a possibly undefined parameter before an always defined parameter seems weird to me. Maybe we can switch them?

Also, in what case is `services` undefined? Shouldn't we handle it there.

Reply:
Good point. I made it mandatory and it throws an error when there are no services for a connection record.